docs/functional/architecture/0001-Field_View_Frontend_Design.md

```diff
@@ -87,7 +84,38 @@ This step makes use of the type information on expressions, for example to decid
 ##### Assume FOAST is correct
 As per guiding principle (5), the lowering should not worry if the FOAST is incorrect or invalid. It is the responsibility of the previous parser and passes to sanitize the user code.
 
+### Program AST (PAST)
+
+The Program AST is a stateful dialect whose main purpose is to apply operators to fields and storing the result on another field. In the future it could be extended to also support control flow (loops, conditionals).
+
+#### What could be changed
+
+The syntax to call field operators feels unnatural as their signature suggests they return a Field, but instead after decorating they return nothing and have an additional argument `out`.
+
+#### Similarities to FOAST
+
+The PAST is an independent dialect and (code-wise) only shares symbol types with the FOAST. This was a deliberate choice as there is currently no good way to share nodes and we wanted to allow changing each of the dialects independently until they become stable. Nonetheless nodes that exist in both dialects should stay in sync so that they could be generalized in the future.
+
+#### PAST parser
+Parses from Python AST into PAST.
+
+#### PAST Lowering
+Lowers from PAST to Iterator IR.
+
+__Field operator out argument slicing__
+The lowering contains some complex logic to emulate slicing of fields in order to restrict the output domain of a field operator call. This contradics guiding principle (3) and should be removed in the future. The concept of specifying the domain for Field operators was not investigated during the frontend design as we concentrated on the beautified iterator dialect where the output domain is explicitly given when lifting a local operator to a field operator. The alternative to also explicitly specify the domain for calls to field operators was rejected as it:
```

Comment:
I might be wrong but I think this is not completely accurate: a field operator could return multiple fields with different domains, so the main problem to specify the output domain in field view is not the syntax, but the fact that each output field could be defined in a different domain. 

Besides that, I agree that the syntax could be further discussed and improved. For example, an alternative strategy just thought about right now could be to use a special marker to allow in place updates of the fields with assignments syntax (a bit like `...` is used in NumPy for a full replace of the array content)

```python
out_a[IN_PLACE], out_b[IN_PLACE] = field_op(a, b)
```

Reply:
With the current implementation of the iterator IR a stencil closure has a single domain and as such a field operator when called at the program level can only return fields with a single output domain. This is one of my critic points of the model.

With respect to the syntax, what speaks against using `...`?

Reply:
> With the current implementation of the iterator IR a stencil closure has a single domain and as such a field operator when called at the program level can only return fields with a single output domain. This is one of my critic points of the model.

You're right about the current implementation of the lowering from field view, although theoretically I guess it could be done differently to support different output domains per field in a (hypothetical) future.

> With respect to the syntax, what speaks against using `...`? 

I would argue that the accepted meaning of `...` is overwrite the **full** content of the field and we want something to indicate that only the matching domain should be update. Anyway, this is not important for now.

Reply:
Ah, I see what you mean with the `IN_PLACE` now. I would argue that the domain as specified using the slice and the domain as given by the field returned from the field operator need to agree and that there is no real need to inherit the domain from the returned field (the same way as this is not needed in numpy). Having `...` overwrite the full content of the field is what I intended it to mean and I believe this is the common case for unstructured and also for cartesian if you handle the boundary conditions inside a field operator.